wandb/magic_impl.py

```diff
@@ -0,0 +1,354 @@
+                        },
+                    'wandb': {
+                        'enable': True
+                        }
+                    }
+                }
+            }
+        }
+    }
+            
+
+def _parse_magic(val):
+    # attempt to treat string as a json
+    if val is None:
+        return _magic_defaults
+    if val.startswith("{"):
+        try:
+            val = json.loads(val)
+        except ValueError:
+            wandb.termwarn("Unable to parse magic json", repeat=False)
+            return _magic_defaults
+        return _merge_dicts(_magic_defaults, val)
+    if os.path.isfile(val):
+        try:
+            with open(val, 'r') as stream:
+                val = yaml.safe_load(stream)
+        except IOError as e:
+            wandb.termwarn("Unable to read magic config file", repeat=False)
+            return _magic_defaults
+        except yaml.YAMLError as e:
+            wandb.termwarn("Unable to parse magic yaml file", repeat=False)
+            return _magic_defaults
+        return _merge_dicts(_magic_defaults, val)
+    # parse as a list of key value pairs
+    if val.find('=') > 0:
+        conf = {}
+        _merge_dicts(_magic_defaults, conf)
+        for kv in val.split(','):
+            kv = kv.split('=')
+            if len(kv) != 2:
+                wandb.termwarn("Unable to parse magic key value pair", repeat=False)
+                continue
+            d = _dict_from_keyval(*kv)
+            _merge_dicts(d, conf)
+        return conf
+    wandb.termwarn("Unable to parse magic parameter", repeat=False)
+    return _magic_defaults
+        
+
+def set_entity(value, env=None):
+    if env is None:
+        env = os.environ
+
+
+def _fit_wrapper(fn, generator=None, *args, **kwargs):
+    trigger.call('on_fit')
+    keras = sys.modules.get("keras")
+    epochs = kwargs.pop("epochs", None)
+
+    magic_epochs = _magic_get_config("magic.keras.fit.epochs", None)
+    if magic_epochs is not None:
+        epochs = magic_epochs
+    callbacks = kwargs.pop("callbacks", [])
+
+    tb_enabled = _magic_get_config("magic.keras.fit.callbacks.tensorboard.enable", None)
+    if tb_enabled:
+        if not any([isinstance(cb, keras.callbacks.TensorBoard) for cb in callbacks]):
+            tb_callback = keras.callbacks.TensorBoard(log_dir=wandb.run.dir)
+            callbacks.append(tb_callback)
+    
+    wandb_enabled = _magic_get_config("magic.keras.fit.callbacks.wandb.enable", None)
+    if wandb_enabled:
+        if not any([isinstance(cb, wandb.keras.WandbCallback) for cb in callbacks]):
+            wandb_callback = wandb.keras.WandbCallback()
+            callbacks.append(wandb_callback)
+
+    kwargs["callbacks"] = callbacks
+    if epochs is not None:
+        kwargs["epochs"] = epochs
+    if generator:
+        return fn(generator, *args, **kwargs)
+    return fn(*args, **kwargs)
+
+
+# NOTE(jhr): need to spell out all useable args so that users who inspect can see args
```

Comment:
You can probably get the Keras implementation's list of args by inspection.

Reply:
Tricky,  the issue is also that I am trying to fool inspection,  since it is python you are probably right that the function signature can be built based on inspection.  I will look into that on the tf.keras release to see if that is the easiest way to handle things if keras 1, keras 2, and tf.keras variants vary too much, otherwise I was thinking just taking what was common between them and inspection of the wrapper will miss new args